proto/usr/share/initramfs-tools/scripts/live-bottom/triton-networking

```diff
@@ -0,0 +1,178 @@
+    fi
+fi
+
+# Set domain name
+datacenter_name=$(jq -r '.datacenter_name // empty' "${triton_networking_file}")
+dns_domain=$(jq -r '.dns_domain // empty' "${triton_networking_file}")
+if [ "${dns_domain}" != "" ] && [ "${datacenter_name}" != "" ]; then
+    sed -i "s/#Domains=/Domains=${datacenter_name}.${dns_domain}/" "${rootmnt}/etc/systemd/resolved.conf"
+    echo "domain ${datacenter_name}.${dns_domain}" >> "/etc/resolv.conf"
+fi
+
+# Set resolvers for systemd to use
+resolvers=$(jq -r '.resolvers | join(" ")' "${triton_networking_file}")
+if [ "${resolvers}" != "" ]; then
+    sed -i "s/#DNS=/DNS=${resolvers}/" "${rootmnt}/etc/systemd/resolved.conf"
+    # Add resolvers for the current initramfs too:
+    echo "nameserver ${resolvers}" >> "/etc/resolv.conf"
+fi
+
+# Set NTP to use headnode
+ntp_hosts=$(get_config_value ntp_hosts)
+if [ "$ntp_hosts" != "" ]; then
+    sed -i "s/#NTP=/NTP=${ntp_hosts}/" "${rootmnt}/etc/systemd/timesyncd.conf"
+fi
+
+# Create network devices
+counter=0
+while IFS="$(printf '\t')" read -r mac nictag mtu ip netmask gateway resolvers; do
+    counter=$((counter+1))
+    filename="${rootmnt}/etc/systemd/network/0${counter}-${nictag}.network"
+    # shellcheck disable=SC2086
+    cat << EOF >> "${filename}"
+[Match]
+MACAddress=${mac}
+
+[Network]
+Address=${ip}$(IPprefix_by_netmask ${netmask})
+EOF
+
+    # Add DNS resolvers - they are comma separated, e.g. "8.8.8.8,8.8.4.4"
+    dns_resolvers=$(get_config_value dns_resolvers)
+    if [ "$resolvers" != "" ] && [ "$dns_resolvers" != "" ]; then
+        for dns in $(echo "${resolvers},${dns_resolvers}" | sed "s/,/ /g"); do
+            echo "DNS=${dns}" >> "${filename}"
+        done
+    fi
+    # Disable DNSSEC
+    echo "DNSSEC=false" >> "${filename}"
+
+    # Ensure we use triton resolvers to resolve triton domain
+    if [ "$nictag" = "$admin_nic_tag" ]; then
+            echo "Domains=${datacenter_name}.${dns_domain}" >> "${filename}"
+    fi
+
+    if [ "$gateway" != "null" ]; then
+        echo "[Route]" >> "${filename}"
+        echo "Gateway=${gateway}" >> "${filename}"
+    fi
+
+    # Add MTU
+    if [ "$mtu" != "null" ]; then
+        printf "\n[Link]\n" >> "${filename}"
+        echo "MTUBytes=${mtu}" >> "${filename}"
+    fi
+
+    # Create .link file to rename this device
+    linkfilename="${rootmnt}/etc/systemd/network/0${counter}-${nictag}.link"
+    cat <<EOF >> "${linkfilename}"
+[Match]
+MACAddress=${mac}
+
+[Link]
+Name=${nictag}
+EOF
+
+    if [ "${nictag}" = "${admin_nic_tag}" ]; then
+        # The admin (iPXE) nic cannot be renamed whilst it is in use, so we
+        # must manually stop the nic, rename it and then restart it.
+        linkname=$(get_device_by_hwaddr "${mac}")
+        if [ "${linkname}" != "" ]; then
+            # Manually set the device name.
+            ip link set "${linkname}" down
+            ip link set "${linkname}" name "${nictag}"
+            ip link set "${nictag}" up
+        fi
+    fi
+
+done <<EOJQ
```

Comment:
Closing the while loop started on line 112. It's a little funky because the indenting isn't consistent due to the numerous HERE docs used in this loop.

Basically what we're doing is constructing a config file per interface, and each interface has a bunch of conditional items. This is something that would be ripe for a nice templating language, but we're extremely early in boot here and not even all the filesystems are mounted yet so what we've got available is very limited.

Reply:
ok!